src/python/pants/backend/docker/subsystems/dockerfile_wrapper_script.py

```diff
@@ -28,7 +28,22 @@ class ParsedDockerfileInfo:
 
 _address_regexp = re.compile(
     r"""
-    (?://)?[^:# ]*:[^:#!@?/\= ]+(?:\#[^:#!@?= ]+)?$
+    # Optionally root:ed.
+    (?://)?
+    # Optional path.
+    [^:# ]*
+    # Optional target name.
+    (?::[^:#!@?/\= ]+)?
```

Comment:
It looks like this was previously required, but is now optional. Is that an intentional change?

Reply:
yes, I looked at https://github.com/pantsbuild/pants/blob/c15e54e5c8100a7d8638589bc6f675a95e7c49e9/src/rust/engine/address/src/lib.rs#L45-L46 when re-modeling this to cover an actual valid address as close as possible.

Reply:
Ah, okay. Given everything is now optional, I wonder if this means dependency inference will now be looking at ~every bit of text in a docker file, e.g. random text like `abc` matches the "optional path" bit (and, even an empty string now satisfies this regex, I think).

The change to allow inferring without a target name seems somewhat of separate to supporting parametrised targets, and is potentially more "dangerous"? For instance, being more likely to change user-observed behaviour by inferring spurious dependencies or performance issues.

Some options I can think of:

1. make the change here and cherrypick it all, as originally planned
2. make the change here but avoid cherrypicking
3. do the two changes separately, and cherry pick the support for parameters but not the optional-name one
4. something else?

Thoughts?

Reply:
Astute observations, as always :) 

I think, trying to recollect the original train of thought as best I can, could've been to be a little conservative with what values to pick up as potential addresses to infer as dependencies to, as back then the "try this value and see if it resolves to something" was a little ugly. Now when it's explicitly OK to [not fail](https://github.com/pantsbuild/pants/blob/af1b5c7cc6f8f539be353b5f950c9d4c36a49138/src/python/pants/backend/docker/util_rules/dependencies.py#L60) for invalid values (not an actual existing address) but merely doesn't resolve to an address, I don't see any reason to not try every value that could remotely resemble a valid address.

The thing is, the values we do try to use are from a very specific narrow part of the whole Dockerfile, namely _default build arg values that is also used in the FROM instruction_. So, any build arg that isn't used in the FROM instruction, won't be considered. Any build arg that doesn't have a default value.. same. etc. So, given that I don't think this is going to hit performance noticeably, and we shouldn't get any false hits either. (I could make a case up where this is not true, but I don't think we'll see that in practice.)

So I'm leaning towards option 1. How does that sound?

Reply:
> > The thing is, the values we do try to use are from a very specific narrow part of the whole Dockerfile, namely default build arg values that is also used in the FROM instruction. So, any build arg that isn't used in the FROM instruction, won't be considered.
> 
> Ah okay, cool. To confirm, does it also apply to `FROM //some/path/to:target`?

No, it does not, as Pants doesn't patch/edit the Dockerfile, so there would not be a way to replace that value with the actual image name.